### PR TITLE
Purecap kernel changes to atomics headers

### PR DESCRIPTION
**Pull request by qwattash**

### CODE REVIEW

File: sys/mips/include/atomic.h [hunk 1]
```diff
@@ -555,6 +559,141 @@ ATOMIC_STORE_LOAD(64)
 #define	atomic_load_64	atomic_load_acq_64
 #endif
 
+#ifdef __CHERI_PURE_CAPABILITY__
+/*
+ * In a purecap kernel we can not use the generic sub-word implementation
+ * as it will break with subobject bounds.
```
Comment (jrtc27) [inline sys/mips/include/atomic.h:565]: Breaks even without subobject bounds

File: sys/mips/include/atomic.h [hunk 2]
```diff
@@ -555,6 +559,141 @@ ATOMIC_STORE_LOAD(64)
 #define	atomic_load_64	atomic_load_acq_64
 #endif
 
+#ifdef __CHERI_PURE_CAPABILITY__
+/*
+ * In a purecap kernel we can not use the generic sub-word implementation
```
Comment (jrtc27) [inline sys/mips/include/atomic.h:564]:
```suggestion
 * In a purecap kernel we cannot use the generic sub-word implementation
```

File: sys/mips/include/atomic.h [hunk 3]
```diff
@@ -882,6 +1021,229 @@ atomic_fetchadd_64(__volatile uint64_t *p, uint64_t v)
 }
 #endif
 
+
+#ifdef __CHERI_PURE_CAPABILITY__
+/*
+ * cheri-only variants to perform atomic operations on pointers.
```
Comment (jrtc27) [inline sys/mips/include/atomic.h:1027]:
```suggestion
 * CHERI-only variants to perform atomic operations on pointers.
```

File: sys/mips/include/atomic.h [hunk 4]
```diff
@@ -882,6 +1021,229 @@ atomic_fetchadd_64(__volatile uint64_t *p, uint64_t v)
...
+ */
+
+static __inline void
+atomic_set_cap(__volatile uintptr_t *p, uintptr_t v)
```
Comment (jrtc27) [inline sys/mips/include/atomic.h:1031]: Use _ptr or a uintcap_t type, but don't mix things like this (I prefer the former, as is done on RISC-V).

File: sys/riscv/include/atomic.h [hunk 5]
```diff
@@ -322,13 +348,114 @@ atomic_store_rel_64(volatile uint64_t *p, uint64_t val)
...
+
+	__asm __volatile(
+		"1:	clr.c	%1, %0\n"
+		"	cgetaddr %2, %1\n"
+		"	or	%2, %2, %3\n"
+		"	csetaddr %1, %1, %2\n"
+		"	csc.c	%2, %1, %0\n"
+		"	bnez	%2, 1b\n"
+		: "+A" (*p), "=&C" (temp1), "=&r" (temp2)
+		    : "r" ((vaddr_t)val)
```
Comment (jrtc27) [inline sys/riscv/include/atomic.h:395]: ptraddr_t (repeated below), plus indentation is strange

File: sys/mips/include/atomic.h [hunk 6]
```diff
@@ -90,8 +90,11 @@ static __inline int atomic_fcmpset_8(__volatile uint8_t *, uint8_t *, uint8_t);
 static __inline int atomic_cmpset_16(__volatile uint16_t *, uint16_t, uint16_t);
 static __inline int atomic_fcmpset_16(__volatile uint16_t *, uint16_t *, uint16_t);
 
-static __inline void
 /* Work around https://github.com/CTSRD-CHERI/qemu/issues/4 */
+#define QEMU_TLB_WORKAROUND8(register) \
```
Comment (jrtc27) [inline sys/mips/include/atomic.h:94]: Do we still need these? Don't really want to be adding more workarounds, and I think the existing ones should be long overdue for removal.

Comment (arichardson) [reply sys/mips/include/atomic.h:94]: Yeah these should go. No longer required anymore because we fixed QEMU to return to the correct address on trap a long time ago.

File: sys/mips/include/atomic.h (hunk 4 above)
Comment (qwattash) [reply sys/mips/include/atomic.h:1031]: Done the former, mirroring RISC-V.

File: sys/mips/include/atomic.h [hunk 7]
```diff
@@ -90,20 +90,14 @@ static __inline int atomic_fcmpset_8(__volatile uint8_t *, uint8_t *, uint8_t);
 static __inline int atomic_cmpset_16(__volatile uint16_t *, uint16_t, uint16_t);
 static __inline int atomic_fcmpset_16(__volatile uint16_t *, uint16_t *, uint16_t);
 
-static __inline void
```
Comment (arichardson) [inline sys/mips/include/atomic.h:93]: I'd split the QEMU_TLB_WORKAROUND32 removal out into another commit

File: sys/mips/include/atomic.h [hunk 8]
```diff
@@ -882,6 +992,229 @@ atomic_fetchadd_64(__volatile uint64_t *p, uint64_t v)
...
+
+static __inline void
+atomic_set_ptr(volatile uintptr_t *p, uintptr_t v)
+{
+	uintptr_t temp;
+
+	__asm __volatile (
+		"1:\n\t"
+		"cllc		%0, %1\n\t"	/* load old value */
+		"cgetoffset	$t0, %0\n\t"	/* calculate new value */
```
Comment (arichardson) [inline sys/mips/include/atomic.h:1009]: Shouldn't we be using the address here?

File: sys/mips/include/atomic.h [hunk 9]
```diff
@@ -882,6 +992,229 @@ atomic_fetchadd_64(__volatile uint64_t *p, uint64_t v)
...
+static __inline void
+atomic_clear_ptr(volatile uintptr_t *p, uintptr_t v)
+{
+	uintptr_t temp;
+	v = ~v;
+
+	__asm __volatile (
+		"1:\n\t"
+		"cllc		%0, %1\n\t"	/* load old value */
+		"cgetoffset	$t0, %0\n\t"	/* calculate new value */
```
Comment (arichardson) [inline sys/mips/include/atomic.h:1030]: same

File: sys/mips/include/atomic.h [hunk 10]
```diff
@@ -882,6 +992,229 @@ atomic_fetchadd_64(__volatile uint64_t *p, uint64_t v)
...
+
+static __inline void
+atomic_add_ptr(volatile uintptr_t *p, uintptr_t v)
+{
+	uintptr_t temp;
+
+	__asm __volatile (
+		"1:\n\t"
+		"cllc		%0, %1\n\t"	/* load old value */
+		"cgetoffset	$t0, %2\n\t"	/* calculate new value */
```
Comment (arichardson) [inline sys/mips/include/atomic.h:1050]: I'd go for addr here even though it shouldn't make a difference (you'd hope v is NULL-derived).

File: sys/mips/include/atomic.h [hunk 11]
```diff
@@ -555,6 +537,138 @@ ATOMIC_STORE_LOAD(64)
 #define	atomic_load_64	atomic_load_acq_64
 #endif
 
+#ifdef __CHERI_PURE_CAPABILITY__
+/*
+ * In a purecap kernel we cannot use the generic sub-word implementation.
+ */
+static __inline int
```
Comment (arichardson) [inline sys/mips/include/atomic.h:544]: Maybe just use the compiler intrinsic like risc-v does? But I guess the inline asm code here exists so we might as well use it.
https://cheri-compiler-explorer.cl.cam.ac.uk/z/48xYza

File: sys/sys/atomic_common.h [hunk 12]
```diff
@@ -45,7 +45,7 @@
 #define	atomic_load_8(p)	(*(volatile uint8_t *)(p))
 #define	atomic_load_16(p)	(*(volatile uint16_t *)(p))
 #define	atomic_load_32(p)	(*(volatile uint32_t *)(p))
-#ifdef _LP64
```
Comment (arichardson) [inline sys/sys/atomic_common.h:48]: I should really update the compiler to poison `_LP64/__LP64__` in purecap mode, but last time I had a look that also reported errors even if the identifier was not actually being used.

Comment (arichardson) [reply sys/sys/atomic_common.h:48]: https://cheri-compiler-explorer.cl.cam.ac.uk/z/1Panf4

File: sys/riscv/include/atomic.h [hunk 13]
```diff
@@ -354,14 +354,46 @@ static __inline void
...
 atomic_clear_ptr(volatile uintptr_t *p, uintptr_t val)
 {
 
+#ifdef notyet
 	(void)__atomic_and_fetch(p, ~val, __ATOMIC_RELAXED);
+#else
+	uintptr_t temp1;
+	u_long temp2;
+
+	__asm __volatile(
```
Comment (arichardson) [inline sys/riscv/include/atomic.h:410]: We should really do this expansion in the compiler but it's kinda awkward since built-in support for RISC-V expansion only allows expanding it to a cmpxchg sequence.

File: sys/riscv/include/atomic.h [hunk 14]
```diff
@@ -322,13 +324,82 @@ atomic_store_rel_64(volatile uint64_t *p, uint64_t val)
...
 #define	atomic_set_acq_ptr		atomic_set_acq_64
 #define	atomic_subtract_acq_ptr		atomic_subtract_acq_64
+#endif
+
+#ifdef __CHERI_PURE_CAPABILITY__
+static __inline void
+atomic_add_ptr(volatile uintptr_t *p, uintptr_t val)
+{
+
+	(void)__atomic_add_fetch(p, val, __ATOMIC_RELAXED);
```
Comment (arichardson) [inline sys/riscv/include/atomic.h:366]: These won't work (yet) since they result in libcalls.

File: sys/mips/include/atomic.h (hunk 7 above)
Comment (qwattash) [reply sys/mips/include/atomic.h:93]: done

File: sys/mips/include/atomic.h (hunk 8 above)
Comment (qwattash) [reply sys/mips/include/atomic.h:1009]: Yes, this might have been wrong even predating the get/setaddr instructions as the base should have been taken into account.

File: sys/mips/include/atomic.h (hunk 10 above)
Comment (qwattash) [reply sys/mips/include/atomic.h:1050]: Technically this should be the case for `subtract_ptr` as well

File: sys/sys/atomic_common.h (hunk 12 above)
Comment (qwattash) [reply sys/sys/atomic_common.h:48]: That would be nice, I suspect there are still a bunch lurking around.

File: sys/riscv/include/atomic.h (hunk 14 above)
Comment (bsdjhb) [reply sys/riscv/include/atomic.h:366]: Yes, the ones that the kernel actually uses I ended up patching to use inline implementations.

File: sys/riscv/include/atomic.h [hunk 15]
```diff
@@ -104,10 +104,25 @@ atomic_fcmpset##SUFFIX##WIDTH(__volatile uint##WIDTH##_t *p,		\
 	ATOMIC_FCMPSET_ORDER(WIDTH, _, __ATOMIC_RELAXED)		\
 	ATOMIC_FCMPSET_ACQ_REL(WIDTH)					\
 
+#ifdef __CHERI_PURE_CAPABILITY__
+/*
+ * The purecap kernel can not use the generic sub-word atomics
```
Comment (bsdjhb) [inline sys/riscv/include/atomic.h:109]: Should fix this to match the updated MIPS comment.

File: sys/riscv/include/atomic.h [hunk 16]
```diff
@@ -104,10 +104,25 @@ atomic_fcmpset##SUFFIX##WIDTH(__volatile uint##WIDTH##_t *p,		\
...
+ * as they break in the presence of subobject bounds.
+ */
+ATOMIC_CMPSET(8);
+ATOMIC_FCMPSET(8);
+ATOMIC_CMPSET(16);
+ATOMIC_FCMPSET(16);
+#define	atomic_cmpset_8 atomic_cmpset_8
+#define	atomic_fcmpset_8 atomic_fcmpset_8
+#define	atomic_cmpset_16 atomic_cmpset_16
+#define	atomic_fcmpset_16 atomic_fcmpset_16
```
Comment (bsdjhb) [inline sys/riscv/include/atomic.h:119]:
```suggestion
#define	atomic_cmpset_8		atomic_cmpset_8
#define	atomic_fcmpset_8		atomic_fcmpset_8
#define	atomic_cmpset_16		atomic_cmpset_16
#define	atomic_fcmpset_16	atomic_fcmpset_16
```

(Adjust the tabs as needed to line up the column with `atomic_cmpset_char` below)

**Review by bsdjhb — COMMENTED** [review]
I agree with splitting out the change to remove the QEMU workaround as a separate commit before the MIPS purecap changes.

File: sys/mips/include/atomic.h (hunk 11 above)
Comment (qwattash) [reply sys/mips/include/atomic.h:544]: I can use the compiler intrinsics, but if I do I think we have to ditch the assembly for the variants non specific to the purecap kernel as well. It is probably for the best anyway though.

File: sys/riscv/include/atomic.h (hunk 14 above)
Comment (qwattash) [reply sys/riscv/include/atomic.h:366]: This is apparently unused, so perhaps we leave it like this until we either will need an implementation or have compiler support?

Comment (arichardson) [reply sys/riscv/include/atomic.h:365]: Sounds fine to me.

**Review by bsdjhb — APPROVED** [review]
I haven't really looked at the MIPS ones in detail, but the rest looks good to me.

File: sys/mips/include/atomic.h [hunk 17]
```diff
@@ -882,6 +712,232 @@ atomic_fetchadd_64(__volatile uint64_t *p, uint64_t v)
...
+
+ATOMIC_ACQ_REL_PTR(set)
+ATOMIC_ACQ_REL_PTR(clear)
+ATOMIC_ACQ_REL_PTR(add)
+ATOMIC_ACQ_REL_PTR(subtract)
+
+#undef ATOMIC_ACQ_REL_PTR
+
+static __inline int
+atomic_cmpset_ptr(volatile uintptr_t *p, uintptr_t cmpval, uintptr_t newval)
```
Comment (arichardson) [inline sys/mips/include/atomic.h:825]: Can we use 
```
ATOMIC_CMPSET(ptr);
ATOMIC_FCMPSET(ptr);
```
here in the same was as the RISC-V file?
It seems to generate mostly the same code except for some unnecessary shifting: https://cheri-compiler-explorer.cl.cam.ac.uk/z/Y1ofTa

Comment (jrtc27) [reply sys/mips/include/atomic.h:825]: The shifting should be fixed by https://reviews.llvm.org/rG34da12dd1fb327bb56f5003a727a73c78929a98d if we care enough to cherry-pick it.
